File: intrinsic/util/path_resolver/path_resolver.go
// Copyright 2023 Intrinsic Innovation LLC

// Package pathresolver provides functionality to access runfiles
package pathresolver

import (
	"fmt"
	"io/fs"
	"os"
	"path"
	"path/filepath"

	"github.com/bazelbuild/rules_go/go/runfiles"
)

// ResolveRunfilesFs gets an fs.FS for runfiles relative to the repo root.
func ResolveRunfilesFs() (fs.FS, error) {
	r, err := runfiles.New()
	if err != nil {
		return nil, fmt.Errorf("unable to create runfiles object: %v", err)
	}
	baseDirFs, err := fs.Sub(r, "ai_intrinsic_sdks")
	if err != nil {
		return nil, fmt.Errorf("unable to get runfiles sub: %v", err)
	}
	return baseDirFs, nil
}

// ResolveRunfilesPath gets the runfiles location of a file.
//
// Use the typical runfiles path without the repository name.
//
// Correct:
//
//	ResolveRunfilesPath("intrinsic/skills/build_defs/tests/no_op_skill_py_manifest.pbbin")
//
// Incorrect:
//
//	ResolveRunfilesPath("intrinsic/skills/build_defs/tests/no_op_skill_py_manifest.pbbin")
func ResolveRunfilesPath(p string) (string, error) {
	r, err := runfiles.New()
	if err != nil {
		return "", err
	}
	return r.Rlocation(path.Join("ai_intrinsic_sdks", p))
}

// ResolveRunfilesOrLocalPath gets the runfiles or local location of a file.
//
// This is useful when packaging inside a container, where the runfiles directory
// is not available.
func ResolveRunfilesOrLocalPath(p string) (string, error) {
	resolvedPath, errRunfile := ResolveRunfilesPath(p)
	if errRunfile == nil {
		return resolvedPath, nil
	}
	resolvedRepo := "ai_intrinsic_sdks~override"
	resolvedPath = filepath.Join(".", resolvedRepo, p)
	if _, err := os.Stat(resolvedPath); err != nil {
		return "", fmt.Errorf("unable to resolve path %q:\n  Not available as runfile: %v\n  Not available as local file: %v", p, errRunfile, err)
	}
	return resolvedPath, nil
}

File: intrinsic/util/path_resolver/path_resolver.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/util/path_resolver/path_resolver.h"

#include <string>

#include "absl/log/log.h"
#include "absl/strings/match.h"
#include "absl/strings/string_view.h"
#include "ortools/base/path.h"
#include "tools/cpp/runfiles/runfiles.h"

namespace intrinsic {

using bazel::tools::cpp::runfiles::Runfiles;

std::string PathResolver::ResolveRunfilesDir() {
  std::string error;
  auto runfiles = std::unique_ptr<Runfiles>(
      Runfiles::Create(program_invocation_name,  // Needed for b/382028959
                       &error));
  if (runfiles == nullptr) {
    LOG(ERROR) << "Error creating Runfiles object: " << error;
    return "";
  }
  return runfiles->Rlocation("ai_intrinsic_sdks~");
}

std::string PathResolver::ResolveRunfilesPath(absl::string_view path) {
  return file::JoinPath(ResolveRunfilesDir(), path);
}

std::string PathResolver::ResolveRunfilesPathForTest(absl::string_view path) {
  std::string error;
  auto runfiles = Runfiles::CreateForTest(BAZEL_CURRENT_REPOSITORY, &error);
  if (runfiles == nullptr) {
    LOG(ERROR) << "Error creating Runfiles object for test: " << error;
    return "";
  }
  std::string runfiles_dir = runfiles->Rlocation(BAZEL_CURRENT_REPOSITORY);
  return file::JoinPath(runfiles_dir, path);
}

std::string PathResolver::ResolveRunfilesPathIfRelative(
    absl::string_view path) {
  if (absl::StartsWith(path, "/")) return std::string(path);
  return ResolveRunfilesPath(path);
}

}  // namespace intrinsic
